File: app/cli/hitl_batch.py
"""统一 hitl_required SSE 展开为 Client 队列项（对齐 Go client/internal/hitl/hitl_batch.go）。"""

from __future__ import annotations

from typing import Any

HITL_TYPE_USER_INFORMATION = "user_information"
HITL_TYPE_EXECUTE_TOOL = "execute_tool"


def _hitl_items_from_data(raw: Any) -> list[dict[str, Any]]:
    if not isinstance(raw, list):
        return []
    out: list[dict[str, Any]] = []
    for item in raw:
        if isinstance(item, dict):
            out.append(item)
    return out


def user_information_data_from_hitl_item(item: dict[str, Any]) -> dict[str, Any] | None:
    if not item:
        return None
    data: dict[str, Any] = {"display_type": "normal_text"}
    content = str(item.get("content") or "").strip()
    if content:
        data["content"] = content
    args = item.get("user_information_args")
    if isinstance(args, dict):
        data["user_information_args"] = args
    return data


def _hitl_routing_fields_from_batch(batch: dict[str, Any]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    child_id = str(batch.get("child_session_id") or "").strip()
    if child_id:
        out["child_session_id"] = child_id
    scope = str(batch.get("hitl_scope") or "").strip()
    if scope:
        out["hitl_scope"] = scope
    purpose = str(batch.get("child_purpose") or "").strip()
    if purpose:
        out["child_purpose"] = purpose
    return out


def approval_data_from_hitl_batch(batch: dict[str, Any], execute_items: list[dict[str, Any]]) -> dict[str, Any] | None:
    if not execute_items:
        return None
    data: dict[str, Any] = {
        "approval_type": "execute_tool",
        "approval_args": {"tool_calls": list(execute_items)},
        "display_type": "normal_text",
        **_hitl_routing_fields_from_batch(batch),
    }
    hitl_id = str(batch.get("hitl_id") or "").strip()
    if hitl_id:
        data["approval_id"] = hitl_id
    message = str(batch.get("message") or "").strip()
    if message:
        data["message"] = message
    return data


def expand_hitl_required(data: dict[str, Any]) -> tuple[list[dict[str, Any]], dict[str, Any] | None]:
    """将 hitl_required 展开为 (user_information 队列项, approval 队列项)。"""
    routing = _hitl_routing_fields_from_batch(data)
    user_infos: list[dict[str, Any]] = []
    execute_items: list[dict[str, Any]] = []
    for item in _hitl_items_from_data(data.get("items")):
        hitl_type = str(item.get("hitl_type") or "").strip()
        if hitl_type == HITL_TYPE_USER_INFORMATION:
            ui = user_information_data_from_hitl_item(item)
            if ui:
                user_infos.append({**ui, **routing})
        elif hitl_type == HITL_TYPE_EXECUTE_TOOL:
            execute_items.append(item)
    approval = approval_data_from_hitl_batch(data, execute_items)
    return user_infos, approval

File: client/internal/hitl/hitl_batch.go
package hitl

import (
	"fmt"
	"strings"
)

const (
	HITLTypeUserInformation = "user_information"
	HITLTypeExecuteTool     = "execute_tool"
)

// UserInformationDataFromHITLItem 从 hitl_required item 构造 user_information_required 形 SSE data。
func UserInformationDataFromHITLItem(item map[string]any) map[string]any {
	if item == nil {
		return nil
	}
	data := map[string]any{
		"display_type": "normal_text",
	}
	if content, ok := item["content"].(string); ok && strings.TrimSpace(content) != "" {
		data["content"] = content
	}
	if args, ok := item["user_information_args"].(map[string]any); ok {
		data["user_information_args"] = args
	}
	return data
}

// ApprovalDataFromHITLBatch 从 hitl_required 与 execute_tool items 构造 approval_required 形 SSE data。
func ApprovalDataFromHITLBatch(batch map[string]any, executeItems []map[string]any) map[string]any {
	if len(executeItems) == 0 {
		return nil
	}
	toolCalls := make([]any, 0, len(executeItems))
	for _, item := range executeItems {
		toolCalls = append(toolCalls, item)
	}
	data := map[string]any{
		"approval_type": "execute_tool",
		"approval_args": map[string]any{"tool_calls": toolCalls},
		"display_type":  "normal_text",
	}
	copyHitlRoutingFields(batch, data)
	if hitlID := strings.TrimSpace(fmt.Sprint(batch["hitl_id"])); hitlID != "" && hitlID != "<nil>" {
		data["approval_id"] = hitlID
	}
	if msg := strings.TrimSpace(fmt.Sprint(batch["message"])); msg != "" && msg != "<nil>" {
		data["message"] = msg
	}
	return data
}

func copyHitlRoutingFields(batch map[string]any, dst map[string]any) {
	if batch == nil || dst == nil {
		return
	}
	if id := ChildSessionIDFromData(batch); id != "" {
		dst["child_session_id"] = id
	}
	if scope := strings.TrimSpace(fmt.Sprint(batch["hitl_scope"])); scope != "" && scope != "<nil>" {
		dst["hitl_scope"] = scope
	}
	if purpose := strings.TrimSpace(fmt.Sprint(batch["child_purpose"])); purpose != "" && purpose != "<nil>" {
		dst["child_purpose"] = purpose
	}
}

func hitlItemsFromData(raw any) []map[string]any {
	switch items := raw.(type) {
	case []any:
		out := make([]map[string]any, 0, len(items))
		for _, item := range items {
			if m, ok := item.(map[string]any); ok {
				out = append(out, m)
			}
		}
		return out
	case []map[string]any:
		return items
	default:
		return nil
	}
}

// ExpandHITLRequired 将 hitl_required SSE 展开为 Client 可入队的 user_information / approval 事件序列。
func ExpandHITLRequired(data map[string]any) (userInfo []map[string]any, approval map[string]any) {
	var executeItems []map[string]any
	for _, item := range hitlItemsFromData(data["items"]) {
		switch strings.TrimSpace(fmt.Sprint(item["hitl_type"])) {
		case HITLTypeUserInformation:
			if ui := UserInformationDataFromHITLItem(item); ui != nil {
				copyHitlRoutingFields(data, ui)
				userInfo = append(userInfo, ui)
			}
		case HITLTypeExecuteTool:
			executeItems = append(executeItems, item)
		}
	}
	if len(executeItems) > 0 {
		approval = ApprovalDataFromHITLBatch(data, executeItems)
	}
	return userInfo, approval
}
